rfcs/2023-rsp-testing-utils.md

```diff
@@ -0,0 +1,113 @@
+
+
+3. Create high level test util classes to handle common interaction patterns
+
+Many of our components have several core user interaction patterns. For a ComboBox or a Picker, a user will usually open the dropdown via a click, look for a option in the list, and select that option. This is a pretty simple interaction but requires a fair bit of prerequisite knowledge to properly execute in a test. The test writer has to
+find the right element to press, wait for the dropdown to open and transition in, find the listbox and desired option, and finally click on said option and wait for the dropdown to transition out. Sometimes using `user-event` backfires and the order/timing of the underlying events fired doesn't match browser behavior, resulting in a test-env only failing test that is extremely hard to debug.
+
+To solve this, we could create a test helper class for each ARIA pattern (i.e. table, list, select, button, etc) that contains methods for common interactions specific to the pattern as described in [this issue comment](https://github.com/adobe/react-spectrum/issues/3703#issuecomment-1302727331). The interaction methods would need to work for `realTimers` and `fakeTimers` (if using Jest) and would handle all the requisite element lookup and event firing required for the high level interaction pattern. Where ever possible, these interaction methods should use `user-event` or our other interaction utils. Additionally, it may be valuable for each of the classes to look up and expose the internal nodes that users often want to assert against (e.g. The picker's trigger button, label, listbox, etc). A implementation of this could look like the following:
+
+```jsx
+render(
+  <Provider theme={theme}>
+    <Picker label="Test" data-testid="test" onSelectionChange={onSelectionChange} onOpenChange={onOpenChange}>
+      <Item key="one">One</Item>
+      <Item key="two">Two</Item>
+      <Item key="three">Three</Item>
+    </Picker>
+  </Provider>
+);
+
+let picker = new SelectTester({element: screen.getByTestId('test'), timerType: 'real'});
+await picker.selectOption('Three');
+expect(picker.trigger).toHaveTextContent('Three');
+```
+
+Since these classes are specific to each ARIA pattern, they should work for React Aria Components (RAC) as well as React Spectrum (RSP) components and thus need to handle possible differences in what `element` is provided. For instance, the `data-testid` passed to a RSP Picker is attached to the trigger button element but the `data-testid` for a RAC Select could be the one attached to the wrapper element instead. Alternatively, we could just standardize what `element` is expected to be provided to each of the util classes (aka make the user pass in the Select's trigger button element).
+
+These util classes could live in the `@react-aria/test-util` package and re-exported from `react-aria-components`.
```

Comment:
Not sure about re-exporting it from the monopackages. I wouldn't want test utils to accidentally get bundled into people's production apps. I think using the individual package for test utils should be fine.